src/Type/Php/ArrayFilterFunctionReturnTypeReturnTypeExtension.php

```diff
@@ -52,4 +65,42 @@ public function getTypeFromFunctionCall(FunctionReflection $functionReflection,
 		return new ArrayType($keyType, $itemType, true);
 	}
 
+	private function removeFalsey(Type $type): Type
+	{
+		$falseyTypes = new UnionType([
+			new NullType(),
+			new ConstantBooleanType(false),
+			new ConstantIntegerType(0),
+			new ConstantFloatType(0.0),
+			new ConstantStringType(''),
+			new ConstantArrayType([], []),
+		]);
+
+		if ($type instanceof ConstantArrayType) {
+			$keys = $type->getKeyTypes();
+			$values = $type->getValueTypes();
+
+			foreach ($values as $offset => $value) {
+				if (!$falseyTypes->isSuperTypeOf($value)->yes()) {
```

Comment:
What if some value is false|int? The filtering should leave only `int`. So maybe there should be something with `TypeCombinator::remove` instead as well?

Reply:
I'm not sure, what would be right approach in that case.

Correct behaviour would be union of arrays with all combinations, which could result in exponentional amount of arrays in union. Not optimal IMO:
```php
$a = foo(); // array(bool, bool)

array_filter($a); // array(0 => true)|array(1 => true)|array(true, true)|array()
```

Maybe we could generalize array in that case? Or IDK :shrug:

Reply:
No, I was talking about this:

```php
$a = ['a' => 1];
if (rand(0, 1) === 1) {
    $a['b'] = false;
}

// $a is `array('a' => 1)|array('a' => 1, 'b' => false)` which is described by UnionType as array('a' => 1, ?'b' => false)`

This union does not pass `$type instanceof ConstantArrayType` so therefore is not changed by this extension. But it should The result after `array_filter($a)` should be `array('a' => 1)` again.

I just realized how to do it - you should run the algorithm on each array from the union separately and then pass them to `TypeCombinator::union(...$arrays)` again. It will work even if you have just one array.

Reply:
That's your other comment, which I agree with.

This is response to
> What if some value is false|int? The filtering should leave only int. So maybe there should be something with TypeCombinator::remove instead as well?

See the sniper I posted :)

Reply:
I think I get it - every time the `isSuperTypeOf` is `yes`, we can safely remove the offset. Every time it's `no`, we have to leave it intact. And when it's `maybe`, we'd have to do the combinatorial explosion, which is not manageable. So what about, every time it's `maybe`, we'd generalize the constant array? So `array(bool, bool)` becomes `array<true>`, which works. Even with `int|bool` values, we have to generalize it to `array<int|true>`, because the `int` might be `0`. Do you think this works?

Reply:
Additionally, we can remove key types with definitely removed values from the generalized array. So if `'a'` key is `false`, the `'a'` key type won't be in the generalized array.

Reply:
:+1: I think that's probably the best way to handle it.